app/controllers/users/omniauth_callbacks_controller.rb

```diff
@@ -0,0 +1,17 @@
+class Users::OmniauthCallbacksController < Devise::OmniauthCallbacksController
+  def hpi
+    @user = User.from_omniauth(request.env['omniauth.auth'])
+
+    if @user.persisted?
+      sign_in_and_redirect @user, event: :authentication
+    else
+      flash[:danger] = 'Login failed'
+      redirect_to root_path
+    end
+  end
+
+  def failure
```

Comment:
why do we not use this method in the else block above? is this even a method?

Reply:
it is a method which will be called when something in the omniauth login fails. So it does a semantically different thing (omniauth login process failed - something was not ok with openid or so vs persisting the user failed - the server provided malformed data or we did something wrong when reading an existing user)